verify if some objects are null

sc0w opened the pull request.

sc0w committed:
verify if some objects are null

gonosztopi commented on src/ExternalConnector.cpp:
```diff
@@ -244,7 +244,7 @@ char *command_completion(const char *text, int state)
 
 	wxString test(wxString(text, *wxConvCurrent).Lower());
-	while (nextCommand != curCommands->end()) {
+	while (curCommands && (nextCommand != curCommands->end())) {
```
The documentation for `rl_completion_matches` says that _entry_func_ (i.e. `command_completion()` in this case) will first be called with _state_ being zero on the first call, and non-zero on subsequent calls. Thus `curCommands` (and `nextCommand`, too) will always be initialized and non-zero on calls where _state_ is non-zero.

gonosztopi commented on src/KnownFile.cpp:
```diff
@@ -1153,7 +1153,7 @@ void CKnownFile::CreateOfferedFilePacket(
 				}
 			} else {
-				if (!pClient->SupportsLargeFiles()) {
+				if (pClient && (!pClient->SupportsLargeFiles())) {
```
The `wxASSERT` call on line 1093 properly checks for both `pClient` and `pServer` being non-null. However, only in debug builds (wxDEBUG enabled). That should be changed to something that works also in release builds (like `wxCHECK_RET`) and further checks can be omitted.

gonosztopi commented on src/SharedFileList.cpp:
```diff
@@ -745,7 +745,7 @@ void CSharedFileList::SendListToServer(){
 	//   - if the compressed size is still >= the original size, we send the uncompressed packet
 	// therefor we always try to compress the packet
-	if (server->GetTCPFlags() & SRV_TCPFLG_COMPRESSION){
+	if (server && (server->GetTCPFlags() & SRV_TCPFLG_COMPRESSION)){
```
First of all, this function should only be called when we're connected to a server. But just in case, if we are not connected (i.e. `GetCurrentServer()` returns `NULL` on line 707) we could bail out immediately, because `SendPacket()` will fail anyway at the end of the function. Of course that would render all further checks for `server` being non-null obsolete.

mrjimenez committed:
Address reviewer feedback on null-pointer guards

ExternalConnector.cpp: remove redundant curCommands null guard; the
else branch already returns NULL when curCommands is NULL, so the
while loop is unreachable with a null pointer.

KnownFile.cpp: replace wxASSERT with wxCHECK_RET so the mutual-
exclusion check on pClient/pServer fires in release builds too, not
only in debug. Drop the redundant pClient null guard added by the PR;
the contract enforced at the function entry is sufficient.

SharedFileList.cpp: bail out early when GetCurrentServer() returns
NULL instead of guarding individual uses of server inline. The inline
null guard added by the PR is removed since it is now covered by the
early return.

mrjimenez commented:
I pushed a follow-up commit ([5cc30c24](https://github.com/sc0w/amule/commit/5cc30c24d)) that addresses @gonosztopi's three inline review comments.

**ExternalConnector.cpp** — The `curCommands &&` guard added by this PR is redundant. The `else` branch in the `state == 0` block already returns `NULL` immediately when `GetSubCommandsFor()` returns null, so the `while` loop is only reachable with a valid `curCommands` pointer. The guard has been removed.

**KnownFile.cpp** — Instead of the inline `pClient &&` guard, the existing `wxASSERT` at the top of `CreateOfferedFilePacket` has been promoted to `wxCHECK_RET`. This enforces the mutual-exclusion contract on `pClient`/`pServer` in release builds as well as debug, which is what the reviewer asked for. The redundant null guard is removed.

**SharedFileList.cpp** — An early return has been added immediately after `GetCurrentServer()` returns null, as suggested. This is cleaner than guarding individual uses of `server` inline, and it also simplifies the `limit` calculation (the ternary is no longer needed). The inline guard added by this PR is removed since it is now covered by the early return.

mrjimenez commented:
The branch has been rebased onto `upstream/master` to resolve merge conflicts. The commit referenced in the previous comment (`5cc30c24`) no longer exists; the equivalent rebased commit is `5417d081`.

mrjimenez commented:
@got3nks , I'd appreciate your review, in case I missed something.

got3nks commented:
@mrjimenez — LGTM in spirit, one trailing cleanup. gonosztopi's review said the early return *"would render all further checks for `server` being non-null obsolete."* The ternary at line 1027 was simplified accordingly, but there's still one stale `server &&` left a few lines down at line 1059:

```cpp
if (!file->IsLargeFile() || (server && server->SupportsLargeFilesTCP())) {
```

After the early return, `server` is guaranteed non-null, so this collapses to:

```cpp
if (!file->IsLargeFile() || server->SupportsLargeFilesTCP()) {
```

Defensive but redundant. Otherwise LGTM — happy to approve once this is in.

mrjimenez commented:
Oops, misread the last line. I will open a new PR for the issue you raised.
